source/src/EventSource.cc

```diff
@@ -0,0 +1,287 @@
+        }
+        
+        if(!iter->second.m_registered)
+        {
+          dqm_warning( "EventSource::sendEvent(): Source '{0}' not registered yet to collector '{1}'. Skipping ...", m_sourceName, collector );
+          continue;
+        }
+        
+        std::string collectCommand = "/dqm4hep/app/evtcol/" + collector + "/collect";
+        m_client.sendCommand(collectCommand, collectBuffer);
+      }  
+    }
+
+    //-------------------------------------------------------------------------------------------------
+
+    void EventSource::getSourceInfo(Json::Value &info)
+    {
+      // source name
+      info["source"] = m_sourceName;
+      
+      // host info
+      StringMap hostInfo; Json::Value hostInfoValue;
+      core::fillHostInfo(hostInfo);
+      
+      for(auto hi : hostInfo)
+        hostInfoValue[hi.first] = hi.second;
+      
+      info["host"] = hostInfoValue;
+      
+      // collectors
+      Json::Value collectorsValue;
+      for(auto colIter : m_collectorInfos)
+        collectorsValue.append(colIter.first);
+      
+      info["collectors"] = collectorsValue;
+      
+      // streamer name
+      info["streamer"] = m_streamerName;
+    }
+    
+    //-------------------------------------------------------------------------------------------------
+    
+    bool EventSource::registerMe(const std::string &collector, const Json::Value &info)
+    {
+      std::string requestName = "/dqm4hep/app/evtcol/" + collector + "/register";
+      Json::StreamWriterBuilder builder;
+      std::unique_ptr<Json::StreamWriter> writer(builder.newStreamWriter());
+      std::ostringstream infoStr;
+      writer->write(info, &infoStr);
+      
+      bool returnValue(false);
+      net::Buffer requestBuffer;
+      auto model = requestBuffer.createModel<std::string>();
+      requestBuffer.setModel(model);
+      model->move(std::move(infoStr.str()));
+      
+      m_client.sendRequest(requestName, requestBuffer, [&returnValue,&collector](const net::Buffer &buffer){
+        Json::Value response;
+        Json::CharReaderBuilder readerBuilder;
+        std::unique_ptr<Json::CharReader> reader(readerBuilder.newCharReader());
+        reader->parse(buffer.begin(), buffer.end(), &response, nullptr);
+        
+        const bool registered(response.get("registered", false).asBool());
+        
+        if(!registered)
+        {
+          const std::string message(response.get("message", "The event collector is maybe not available or is down !").asString());
+          dqm_warning( "EventSource::registerMe(): Couldn't register to source to collector '{0}': {1}", collector, message );
+          returnValue = false;
+        }
+        else
+        {
+          returnValue = true;
+        }
+      });
+      
+      return returnValue;
+    }
+    
+    //-------------------------------------------------------------------------------------------------
+    
+    void EventSource::unregisterMe(const std::string &collector)
```

Comment:
No need to check if source is registered here? 
What happens if it's not and you still try to unregister it? Nothing I guess as you're not parsing any return code within `Client::sendCommand`.

Reply:
No as it is called from destructor. Anyway, as soon as the process that starts an event source is stopped, the server is notified and will also remove the client/event source from the one monitored in the server app.